### PR TITLE
Feature: Add dynamic header, goal, title, and description templates to form designs

### PR DESCRIPTION
**Pull request by jonwaldstein**
## Description

This PR introduces the `header` layout to form designs template api along with `title`, `description`, and `goal`.

In the form builder - when selecting _enable goal_ - the header template along with the goal will be visually displayed in the form preview.  The values reflected are dynamically fetched from the database.

## Affects

**Donation Form Rendering**

If goal is enabled, then the donation form will automatically display a goal progress bar with some basic metrics.  Our classic form design has overloaded the `Goal` and `Header` templates to both add the _secure donation badge_ and the stats above the goal progress.

**Donation Form ViewModel Updates**

We now have a bigger array of data being sent to the front-end through our exports method.  The goal needs a decent amount of data to work and warranted its own DTO called `DonationFormGoalData`.  I wrote tests for all this data and their expected shapes.

Also, I simplified our formSettings conditional logic inside the viewModel by only passing the resulting array merge of form settings in the database and from the request.  We have recently discussed eventually separating the preview and render routes so this will make that a lot easier.

**Observations**

- The header and description text updates are noticeably more slow to preview than other settings

## Visuals

Screen Shot 2022-11-18 at 3 06 01 PM

## Testing Instructions

Pull the branch, run `npm run build` and open a next gen form.
Try out the different settings for Donation Goal, Header, and Description

## Pre-review Checklist

-   [ ] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [x] Includes unit tests
-   [ ] Reviewed by the designer (if follows a design)
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

### CODE REVIEW

**Comment by jonwaldstein** [comment]
@kjohnson okay this beast is ready for review 😌

File: src/NextGen/DonationForm/DataTransferObjects/DonationFormGoalData.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+        return [
+            'type' => $this->goalType->getValue(),
+            'enabled' => $this->isEnabled,
+            'show' => $this->isEnabled,
+            'currentAmount' => $currentAmount,
+            'currentAmountFormatted' => $this->goalType->isAmount() ? Money::fromDecimal(
+                $currentAmount,
+                give_get_currency()
+            )->formatToLocale() : $currentAmount,
```
Comment (JasonTheAdams) [inline src/NextGen/DonationForm/DataTransferObjects/DonationFormGoalData.php:82]: First, you shouldn't use this function without passing the locale. Otherwise, the locale is going to come from the server's locale settings, not the browser.

I saw your comment regarding decimals. Full disclosure, this took me a down a whole rabbit hole as I got curious. This is, no joke, the best way I could find to do it in PHP:
```php
$amount = 12300.45;
$formattedAmount = $fmt->formatCurrency($amount, 'USD');
$pattern = '/(.+)\\' . $decialSymbol . '\d+(.*)/';
$integerAmount = preg_replace($pattern, '${1}${2}', $formattedAmount);
```
That should handle whatever the decimal symbol is as well as not message up left or right-side currency symbols. There just doesn't seem to be a way to tell MoneyPHP or the Intl library to not include the decimal. So interesting. We could add a `$withoutDecimal` parameter to methods within the `CurrencyFacade` class if we felt this was useful. 🤷‍♂️ 

The JS Intl lib has a [formatToParts](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/formatToParts) method which makes this a lot easier on that side.

In short, we may simply be worth passing the amount as an integer along with the currency, and have the front-end format it. That way we get the benefits of the browser locale, and can more easily hide decimal amounts if we want to.

File: src/NextGen/DonationForm/DataTransferObjects/DonationFormViewRouteData.php [hunk 2]
```diff
@@ -34,11 +34,39 @@ public static function fromRequest(array $request): DonationFormViewRouteData
...
+
+            if ($setting === 'true') {
+                return true;
+            }
+
+            if (filter_var($setting, FILTER_VALIDATE_BOOLEAN)) {
+                return (bool)$setting;
+            }
+
+            return $setting;
```
Comment (JasonTheAdams) [inline src/NextGen/DonationForm/DataTransferObjects/DonationFormViewRouteData.php:69]: You're kind of reinventing `filter_var`, here. This whole thing could be:
```php
return filter_var($setting, FILTER_VALIDATE_BOOLEAN);`
```

If you want to prevent false as default on error, you can do:
```php
$filteredValue = filter_var($setting, FILTER_VALIDATE_BOOLEAN, FILTER_NULL_ON_FAILURE);

if ( $filteredValue === null ) {
  throw new InvalidArgumentException('Invalid setting, must be boolean. Received ' . $setting);
}

return $filteredValue;
```

This will handle ["true" and "false" as values](https://www.php.net/manual/en/filter.filters.validate.php).

File: src/NextGen/DonationForm/ViewModels/DonationFormViewModel.php [hunk 3]
```diff
@@ -13,76 +15,124 @@
...
      */
-    private $donationForm;
+    private $donationFormId;
     /**
      * @var BlockCollection
      */
-    private $formBlockOverrides;
+    private $formBlocks;
     /**
      * @var array
```
Comment (JasonTheAdams) [inline src/NextGen/DonationForm/ViewModels/DonationFormViewModel.php:26]: Let's use the `array{designId: int, ...}` syntax here to give us some help.

**Review by JasonTheAdams — CHANGES_REQUESTED** [review]
Lots of great work! I left some notes!

File: src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/form/Header.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+const DescriptionTemplate = getDescriptionTemplate();
+const GoalTemplate = getGoalTemplate();
+
+/**
+ * @unreleased
+ */
+export default function Header() {
+    return (
+        <HeaderTemplate
+            Title={() => (form.settings?.showHeading ? <TitleTemplate text={form.settings.heading} /> : <></>)}
```
[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/form/Header.tsx:18]: Nope, because the template is trying render `<Title />`.  This was a weird one.

File: src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/index.tsx [hunk 5]
```diff
@@ -102,7 +110,12 @@ const template = {
 };
 
 // The following functions are used to retrieve the various templates for the form.
-function getTemplate<NodeProps>(type: string, section: string, htmlTag?: string): FC<NodeProps> {
+function getTemplate<NodeProps>(
+    type: string,
+    section: string,
+    htmlTag?: string,
+    template = activeFormDesignTemplate
```
[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/index.tsx:117]: This is leftover from when I was playing around with bringing in the default `Goal` template so I don't have to re-write the html in classic form.  I ended up just importing it in directly.  I can see something like this being very useful though in practice.

File: src/NextGen/DonationForm/DataTransferObjects/DonationFormGoalData.php (hunk 1 above)
Comment (jonwaldstein) [reply src/NextGen/DonationForm/DataTransferObjects/DonationFormGoalData.php:82]: Yeah another weird one.  I originally had all the amount formatting taking place on the front-end using `Intl` which might still end up making the most sense.  That's why I added that `useCurrencyFormatter` in the first place.

File: src/NextGen/DonationForm/DataTransferObjects/DonationFormViewRouteData.php (hunk 2 above)
Comment (jonwaldstein) [reply src/NextGen/DonationForm/DataTransferObjects/DonationFormViewRouteData.php:69]: I think i'm just going to check specifically for the words `true` and `false` and get rid of filter_var.  It's specific to those and don't want this thinking string numbers are booleans.

File: src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/form/Header.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+                        goalLabel={form.goal.label}
+                        progressPercentage={form.goal.progressPercentage}
+                        currentAmount={form.goal.currentAmount}
+                        currentAmountFormatted={form.goal.currentAmountFormatted}
+                        targetAmount={form.goal.targetAmount}
+                        targetAmountFormatted={form.goal.targetAmountFormatted}
+                        totalRevenue={form.stats.totalRevenue}
+                        totalRevenueFormatted={form.stats.totalRevenueFormatted}
+                        totalNumberOfDonationsOrDonors={form.stats.totalNumberOfDonationsOrDonors}
+                        totalNumberOfDonationsOrDonorsLabel={form.stats.totalNumberOfDonationsOrDonorsLabel}
```
[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/form/Header.tsx:36]: This is the one case where something generic just kept confusing me which is why I made it so specific.  If the goal type is 'amount' it still displays the total number of donations.

File: src/NextGen/DonationForm/Controllers/DonationFormViewController.php [hunk 7]
```diff
@@ -147,4 +153,24 @@ private function enqueueFormScripts(int $formId, string $formDesignId)
...
+
+    /**
+     * @unreleased
+     * @return void
+     */
+    private function enqueueFormStyles(DonationFormViewModel $viewModel)
+    {
+        wp_enqueue_global_styles();
+
+        if ($viewModel->primaryFont()) {
```
[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Controllers/DonationFormViewController.php:165]: we don't actually have a setting for this yet - just wanted to match the classic design.  we can manually check for classic design specifically for now until we decide how we're handling from design specific settings vs form settings and fonts as a whole.

File: src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/index.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/index.tsx:117]: I'd rather not introduce something like this until we have a concrete use case for it.

[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/index.tsx:117]: As a form design, If you're trying to get a template from the window - which one are you expecting? The getTemplate function is using the active template so if you're  overriding one of the template parts and also tried to get the template from the window it would cause a loop of inception.  So actually I think using the default template for the window makes the most sense.

File: src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/layouts/Header.tsx [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import {FC} from 'react';
+
+export interface HeaderProps {
+    Title: FC;
+    Description: FC;
+    Goal: FC;
```
[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/layouts/Header.tsx:6]: Actually, even better would be:
```tsx
import type {Goal as GoalComponent} from './some-place}

{
    Goal: typeof GoalComponent
}
```

File: src/NextGen/DonationForm/DataTransferObjects/DonationFormViewRouteData.php (hunk 2 above)
Comment (JasonTheAdams) [reply src/NextGen/DonationForm/DataTransferObjects/DonationFormViewRouteData.php:69]: Does the system know which settings to anticipate as booleans?

File: src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/index.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/index.tsx:117]: Hmm, I'm not quite following. Let's chat about this on Monday.

File: src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/form/Header.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/form/Header.tsx]: resolved ecc8c71e

File: src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/layouts/Title.tsx [hunk 9]
```diff
@@ -0,0 +1,10 @@
+export interface TitleProps {
+    text: string;
+}
+
+/**
+ * @unreleased
+ */
+export default function Title({text}: TitleProps) {
```
[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/layouts/Title.tsx:8]: I'm actually not sure about the Form prefix for title and description.  Technically this is not the form title because this is a separate setting that lives inside the design settings.  In fact the setting is labeled "heading" so that it's clear it's not the actual form title but just a heading that _can_ be displayed above your form. 🤔

File: src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/layouts/Header.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/layouts/Header.tsx:6]: This would be ideal except once you specify the typeof in props, typescript is going to expect it's required props...

Screen Shot 2022-11-21 at 10 23 29 AM

Same here if we changed the FieldLabel you created to use typeof:

Screen Shot 2022-11-21 at 10 24 25 AM

File: src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/layouts/Title.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (kjohnson) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/layouts/Title.tsx:8]: Yeah... this is **not** the form title, but rather a generic design setting (which makes naming difficult).

File: src/NextGen/DonationForm/ViewModels/DonationFormViewModel.php (hunk 3 above)
Comment (jonwaldstein) [reply src/NextGen/DonationForm/ViewModels/DonationFormViewModel.php]: FYI the `$formSettings` property will eventually be a specific type.  Soon we're going to update `$donationForm->settings` to be a `FormSettings` property type instead of a basic array.  I added _some_ array items to the doc with a note here 9bbb4973

File: src/NextGen/DonationForm/DataTransferObjects/DonationFormViewRouteData.php (hunk 2 above)
Comment (jonwaldstein) [reply src/NextGen/DonationForm/DataTransferObjects/DonationFormViewRouteData.php]: resolved up the chain 04e7e109

File: src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/layouts/Title.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/layouts/Title.tsx:8]: I think @kjohnson's suggestion makes the most sense but as `HeaderTitle` and `HeaderDescription`.  Reason being these components live inside the `Header` component and surfaced in the api within _layouts_ .  Just like the `FieldLabel` layout is prefixed with `Field` because it lives inside the `Field`.

[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/layouts/Title.tsx:8]: Jason: 0
Kyle: 1

File: src/NextGen/DonationForm/DataTransferObjects/DonationFormGoalData.php (hunk 1 above)
Comment (jonwaldstein) [reply src/NextGen/DonationForm/DataTransferObjects/DonationFormGoalData.php]: Reformatted to let front-end handle formatting amounts: 577688ae

File: src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/layouts/Title.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/layouts/Title.tsx]: resolved 619eb9fd

File: src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/index.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/index.tsx:120]: The `getTemplate()` function is locating objects using the merge of the default (ours) and active form designs.  So here is an example of a loop:
 
**Scenario**
- classic design is active and supplies it's own `Goal` component
- inside it's `Goal` it uses `window.givewp.templates.getGoal()` and tries to render it (in this case classic doesn't need to change anything to the underlying structure - just add to it)
- this would keep trying to get `classic.layouts.goal` over and over instead of `default.layouts.goal` which is what it wants.  

```
export default function Goal(props: GoalProps) {
    const {
        targetAmountFormatted,
        totalRevenueFormatted,
        totalNumberOfDonationsOrDonors,
        totalNumberOfDonationsOrDonorsLabel,
    } = props;

    // this would now try to get itself inside of itself
   const GoalTemplate = window.givewp.templates.getGoal();

    return (
        <div style={{width: '100%'}}>
            <FormStats
                totalRevenue={totalRevenueFormatted}
                goalTargetAmount={targetAmountFormatted}
                totalNumberOfDonationsOrDonors={totalNumberOfDonationsOrDonors}
                totalNumberOfDonationsOrDonorsLabel={totalNumberOfDonationsOrDonorsLabel}
            />

           // LOOP HERE
            <GoalTemplate {...props} />
        </div>
    );
}
```

[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/index.tsx:120]: Interesting! Makes sense. Sounds like we need a `getDefaultTemplate` function?

[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/index.tsx:120]: Yeah, so what i'm thinking is 
- Internally we should always be getting the active template (merge of active and default design).  
- Externally (on the window / what we're giving to designs to consume) designs should always be getting the default core template we originally supplied.

[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/index.tsx:120]: That makes sense. The only step we haven't added yet is add-ons, so it will be (in cascading order):
- Default templates
- Add-on tempaltes
- Form Design templates

Obviously the Form Design has access to its own templates, but it will need access to the default and add-on templates. Add-ons will only need access to default templates.

[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/templates/index.tsx:120]: Yup sounds good. We don't need to update that stuff in this PR but should in the near future.

**Comment by jonwaldstein** [comment]
@JasonTheAdams most of your feedback has been addressed with a few unresolved.   Feel free to add any more thoughts to those conversations and/or give this another look through.

File: src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/form/Header.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/form/Header.tsx:54]: It just makes me nervous for future goals. For subscription based goals, for example, I think it would make sense for this to be a subscription count — and `totalNumberofDonationsOrDonorsOrSubscriptions` sounds... robust. 😂

[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/form/Header.tsx:54]: Yeah i'm fine changing it to `totalCountValue` for now - that's basically where I started.  I just changed it because I realized that unless it's the 'amount' goal this value is actually the same as `currentAmount` and kept confusing me.

[the remark this answers is not in the document]
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Blocks/DonationFormBlock/resources/app/form/Header.tsx]: resolved e19f8641

File: src/NextGen/DonationForm/Controllers/DonationFormViewController.php [hunk 10]
```diff
@@ -147,4 +154,25 @@ private function enqueueFormScripts(int $formId, string $formDesignId)
...
+        if ($viewModel->primaryFont() && $viewModel->designId() === ClassicFormDesign::id()) {
+            wp_enqueue_style(
+                'givewp-google-font',
+                'https://fonts.googleapis.com/css?family=' . urlencode(
+                    $viewModel->primaryFont()
+                ) . ':400,500,600,700&display=swap',
+                [],
+                GIVE_VERSION
+            );
+        }
```
Comment (JasonTheAdams) [inline src/NextGen/DonationForm/Controllers/DonationFormViewController.php:176]: @jonwaldstein If we're moving away from the font setting for now, can we not have this hard-coded here? Could the Form Design enqueue it itself somehow?

**Review by JasonTheAdams — CHANGES_REQUESTED** [review]
Just have one last thing, but if we need to leave it for now that's fine.

File: src/NextGen/DonationForm/Controllers/DonationFormViewController.php (hunk 10 above)
Comment (jonwaldstein) [reply src/NextGen/DonationForm/Controllers/DonationFormViewController.php]: I removed the primary font stuff 3d3ac9aa

From our custom CSS conversation, there's nothing stopping a form design to add this to their stylesheet: 
```
@import url('https://fonts.googleapis.com/css2?family=Inter:wght@100;200;300;400;500;600;700;800;900&family=Montserrat:wght@300;400;500;700&display=swap');

body {
   font-family: 'Montserrat';
}
```

For now i'm just setting the font to `system-ui` which can easily be updated in the future using the above route or another approach.
